### DISCUSSION TITLE
Best way to reference typed definitions in libraries.

### DISCUSSION THREAD

**Discussion by martinheidegger** [DefinitelyTyped/DefinitelyTyped#53527] [_Misc]
I want to provide typescript definitions in my own library `bar` (published to npm) and it is using a dependency `foo` which's definitions are provided as `@types`. I can think of a few options, but I am not sure: What is the best way to link the dependency?

1. `import foo from 'foo'` without adding definitions to the package dependencies. A user of my library would need to install `@types/foo` in order for things to work. The advantage of this is that it doesn't bloat the package, allows users to always get the latest definitions. It _does_ come however with obvious disadvantage of having to install the dependency manually and possible version incompatibility.
2. `import foo from 'foo'` with adding `@types/foo` to the package dependencies. This may make it clearer for dependencies to have this, but beside update checks, I need to make sure as a library developer that `foo` didn't add types with updates, that may override/conflict with the `@types/foo`.
3. `import foo from '@types/foo'` and adding `@types/foo` to the package dependencies. This seems to be the safest route. This way we get the correct dependencies and types, however: it is very strict in it's definitions and may not reflect the types from the library

Is there another way I don't see? What do you think is the best way?

**Comment by tpluscode** [comment]
I mostly publish packages without dependency of `@types/foo` and let the users install it themselves.

Otherwise, I would add the dependency as `"@types/foo": "*"`, the same way all type packages reference each other. This is the only way which avoids duplicate resolutions which can produce hard to fix errors during build.

Again, the end project may pin a specific version of `@types/foo` which will override the star range
